src/tests/roc_audio/test_watchdog.cpp

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2017 Mikhail Baranov
+ * Copyright (c) 2017 Victor Gaydov
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <CppUTest/TestHarness.h>
+
+#include "roc_audio/watchdog.h"
+#include "roc_core/buffer_pool.h"
+#include "roc_core/heap_allocator.h"
+
+namespace roc {
+namespace audio {
+
+namespace {
+
+enum {
+    MaxBufSize = 5000,
+
+    SamplesPerPacket = 200,
+    NumCh = 2,
+
+    Timeout = 20
+};
+
+core::HeapAllocator allocator;
+core::BufferPool<sample_t> sample_buffer_pool(allocator, MaxBufSize, 1);
+
+class TestFrameReader : public IReader, public core::NonCopyable<> {
+public:
+    void read(Frame& frame) {
+        memset(frame.samples().data(), 0, frame.samples().size() * sizeof(sample_t));
+    }
+};
+
+TestFrameReader test_reader;
+
+} // namespace
+
+TEST_GROUP(watchdog) {
+    Frame new_frame(size_t sz) {
+        core::Slice<sample_t> samples =
+            new (sample_buffer_pool) core::Buffer<sample_t>(sample_buffer_pool);
+        samples.resize(sz * NumCh);
+        Frame frame(samples);
+        return frame;
+    }
+
+    void check_read(unsigned frame_flags, IReader& reader, bool is_read) {
+        Frame frame = new_frame(SamplesPerPacket);
+        frame.add_flags(frame_flags);
+
+        reader.read(frame);
+
+        if (is_read) {
+            for (size_t n = 0; n < frame.samples().size(); n++) {
+                DOUBLES_EQUAL(0.0f, frame.samples().data()[n], 0);
+            }
+        }
+    }
+};
+
+TEST(watchdog, read) {
+    Watchdog watchdog(test_reader, Timeout);
+
+    check_read(0, watchdog, true);
+
+    CHECK(watchdog.update(Timeout / 4));
+    check_read(0, watchdog, true);
+}
+
+TEST(watchdog, read_timeout) {
```

Comment:
I suggest to enhance `read_timeout` and `read` tests a bit:

* `read_timeout1` - call `update(n++)` `Timeout` times, never call read; the session is terminated
* `read_timeout2` - call `update(Timeout+1)` once, never call `read()`; the session is terminated (this is the current `read_timeout` test)
* `read_no_timeout` - call `update(n++)` `Timeout - 1` times, then call `read()` once, then call `update(n++)` once more; the session is not terminated; repeat this several times